### DISCUSSION TITLE
Null conditional foreach

### DISCUSSION THREAD

**Discussion by JC3** [dotnet/csharplang#4774] [Language Ideas] [answered]
I'm not really sure what the feasibility of this is, but I feel like C# is missing a directly null-safe foreach that can handle null collections.

Boilerplate code like this comes up all the time, usually in one of two forms:

    foreach (var Thing in Things ?? SomeEmptyEnumeration) {
    }

And:

    if (Things != null) {
        foreach (var Thing in Things) {
        }
    }

I'm not sure what the best syntax would be, but I'm kind of thinking just:

    foreach (var Thing in Things?) {
    }

Which seems to parallel the syntax of the existing null-conditional (and null-coalescing to some extent) operators in that the `?` appears directly after the value of interest. Other options include `foreach?` or `in?`, but those both feel awkward, and aren't as... "syntactically bound" to the collection variable.

The effect would be that null collections would be treated like empty collections.

It is definitely syntactic sugar, but at the same time so are the null-conditional and null-coalescing operators; and they're very useful. My assumption is that a null-safe `foreach` would be handy in the same way, with similar motivations.

---

Looking at https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#the-foreach-statement, it seems like most (all?) of the existing resolution rules can remain as-is. 

The current equivalent expanded code is stated to be:

```
{
    E e = ((C)(x)).GetEnumerator();
    try {
        while (e.MoveNext()) {
            V v = (V)(T)e.Current;
            embedded_statement
        }
    }
    finally {
        ... // Dispose e
    }
}
```

For the equivalent null-conditional version, there is a bit of an issue here in that we don't have nullable mutable value types, so e.g. this, while tempting, is a problem (remember `E` could theoretically be a non-nullable value type):

```
    E? e = ((C)(x))?.GetEnumerator();
    ....
        while (e?.MoveNext() ?? false) {
```

And so the only thing I can think of is to just wrap the entire vanilla version in an `if (!(x is null)) { ... }` (or `x is not null` where available), for the null-conditional `foreach`. 

One question I have is, is there *any* chance that `(C)(x)` could throw an exception even in the case where `x` is null? I don't know enough about the language to know the answer to that. If that's possible, then a `_ = (C)(x);` should probably be included in the null-conditional `foreach` equivalent code, e.g.

```
if (x is null) {
    _ = (C)(x);
} else {
    E e = ((C)(x)).GetEnumerator();
    try {
        while (e.MoveNext()) {
            V v = (V)(T)e.Current;
            embedded_statement
        }
    }
    finally {
        ... // Dispose e
    }
}
```

... I think. Maybe. 

---

Anyways I'm curious to hear some thoughts on this, in particular:

- If it's even interesting.
- If it's been proposed before (and possibly rejected).
- Best syntax
- Bad unintended side-effects that I can't think of

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/1081

**Reply by JC3** [reply]
Thanks!! Man... that conversation really degenerated into a lot of "my code style is better than yours"...
